SwiftMonkey/MonkeyXCTestPrivate.swift

```diff
@@ -80,146 +88,192 @@ extension Monkey {
-          of all relative probabilities.
-    */
+     Add an action that generates a rotation gesture
+     at a random screen position over a random angle
+     using the private XCTest API.
+
+     - Parameter weight: The relative probability of this
+     event being generated. Can be any value larger than
+     zero. Probabilities will be normalised to the sum
+     of all relative probabilities.
+     */
     public func addXCTestRotateAction(weight: Double) {
         addAction(weight: weight) { [weak self] in
+            guard self?.isInForeground == true else { return }
+
             let rect = self!.randomRect(sizeFraction: 2)
             let angle = CGFloat(self!.r.randomDouble() * 2 * 3.141592)
 
             let semaphore = DispatchSemaphore(value: 0)
-            self!.sharedXCEventGenerator.rotateInRect(rect, withRotation: angle, velocity: 5, orientation: orientationValue) {
-                semaphore.signal()
+            self!.sharedXCEventGenerator.rotateInRect(rect,
+                                                      withRotation: angle,
+                                                      velocity: 5,
+                                                      orientation: orientationValue) { [weak self] in
+                                                        self?.reactivateApplicationIfNeeded()
+                                                        semaphore.signal()
             }
             semaphore.wait()
         }
     }
 
     /**
-        Add an action that generates a device rotation event
-        using the private XCTest API. Does not currently work!
-
-        - Parameter weight: The relative probability of this
-          event being generated. Can be any value larger than
-          zero. Probabilities will be normalised to the sum
-          of all relative probabilities.
-    */
+     Add an action that generates a device rotation event
+     using the private XCTest API. Does not currently work!
+
+     - Parameter weight: The relative probability of this
+     event being generated. Can be any value larger than
+     zero. Probabilities will be normalised to the sum
+     of all relative probabilities.
+     */
     public func addXCTestOrientationAction(weight: Double) {
         addAction(weight: weight) { [weak self] in
+            guard self?.isInForeground == true else { return }
+
             let orientations: [UIDeviceOrientation] = [
                 .portrait,
                 .portraitUpsideDown,
                 .landscapeLeft,
                 .landscapeRight,
                 .faceUp,
                 .faceDown,
-            ]
+                ]
 
             let index = Int(self!.r.randomUInt32() % UInt32(orientations.count))
             orientationValue = orientations[index]
         }
     }
+
+    private func reactivateApplicationIfNeeded() {
+        if #available(iOS 9.0, *) {
+            DispatchQueue.main.async {
```

Comment:
Is it needed to schedule this check to run on the main thread? If so, should we enforce that all added action run on the main thread as well?

Reply:
All of the XCUI* code must be happening on the main thread